js/src/common/components/Tooltip.tsx

```diff
@@ -0,0 +1,275 @@
+  /**
+   * Tooltip textual content.
+   *
+   * String arrays, like those provided by the translator, will be flattened
+   * into strings.
+   */
+  text: string | string[];
+  /**
+   * Manually show tooltip. `false` will show based on cursor events.
+   *
+   * Default: `false`.
+   */
+  tooltipVisible?: boolean;
+  /**
+   * Whether to show on focus.
+   *
+   * Default: `true`.
+   */
+  showOnFocus?: boolean;
+  /**
+   * Tooltip position around element.
+   *
+   * Default: `'top'`.
+   */
+  position?: 'top' | 'bottom' | 'left' | 'right';
+  /**
+   * Whether HTML content is allowed in the tooltip.
+   *
+   * **Warning:** this is a possible XSS attack vector. This option shouldn't
+   * be used wherever possible, and may not work when we migrate to another
+   * tooltip library. Be prepared for this to break in Flarum stable.
+   *
+   * Default: `false`.
+   *
+   * @deprecated
+   */
+  html?: boolean;
+  /**
+   * Sets the delay between a trigger state occurring and the tooltip appearing
+   * on-screen.
+   *
+   * **Warning:** this option may be removed when switching to another tooltip
+   * library. Be prepared for this to break in Flarum stable.
+   *
+   * Default: `0`.
+   *
+   * @deprecated
+   */
+  delay?: number;
+  /**
+   * Used to disable the warning for passing text to the `title` attribute.
+   *
+   * Tooltip text should be passed to the `text` attribute.
+   */
+  ignoreTitleWarning?: boolean;
+}
+
+/**
+ * The `Tooltip` component is used to create a tooltip for an element. It
+ * requires a single child element to be passed to it. Passing multiple
+ * children or fragments will throw an error.
+ *
+ * You should use this for any tooltips you create to allow for backwards
+ * compatibility when we switch to another tooltip library instead of
+ * Bootstrap tooltips.
+ *
+ * If you need to pass multiple children, surround them with another element,
+ * such as a `<span>` or `<div>`.
+ *
+ * **Note:** this component will overwrite the `title` attribute of the first
+ * child you pass to it, as this is how the current tooltip system works in
+ * Flarum. This shouldn't be an issue if you're using this component correctly.
+ *
+ * @example <caption>Basic usage</caption>
+ *          <Tooltip text="You wish!">
+ *            <Button>
+ *              Click for free money!
+ *            </Button>
+ *          </Tooltip>
+ *
+ * @example <caption>Use of `position` and `showOnFocus` attrs</caption>
+ *          <Tooltip text="Woah! That's cool!" position="bottom" showOnFocus>
+ *            <span>3 replies</span>
+ *          </Tooltip>
+ *
+ * @example <caption>Incorrect usage</caption>
+ *          // This is wrong! Surround the children with a <span> or similar.
+ *          <Tooltip text="This won't work">
+ *            Click
+ *            <a href="/">here</a>
+ *          </Tooltip>
+ */
+export default class Tooltip extends Component<TooltipAttrs> {
+  private firstChild: Mithril.Vnode<any, any> | null = null;
```

Comment:
Can't we just `private firstChild!: Mithril.Vnode<any, any>;`

Reply:
I don't like this as it hides the fact that `firstChild` might not be set for some reason by removing undefined/null from its typings.

Reply:
Yes, but that's only done after we know it won't be null.

Your suggestion would hide the original value (actually undefined, I'd guess) in the original typings, meaning that if we edit the file later and attempt to use the variable before we know it's not null, we could unwittingly introduce issues.